# Second Hand Car price prediction

## Objective

Second Hand Car price prediction based on data from Quickr website

### First importing required libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Reading data available in the form of csv file 

In [2]:
cars = pd.read_csv('quikr_car.csv')

### Checking the data (first five rows)

In [3]:
cars.head()

name   company  year          Price  \
0    Hyundai Santro Xing XO eRLX Euro III   Hyundai  2007         80,000   
1                 Mahindra Jeep CL550 MDI  Mahindra  2006       4,25,000   
2              Maruti Suzuki Alto 800 Vxi    Maruti  2018  Ask For Price   
3  Hyundai Grand i10 Magna 1.2 Kappa VTVT   Hyundai  2014       3,25,000   
4        Ford EcoSport Titanium 1.5L TDCi      Ford  2014       5,75,000   

   kms_driven fuel_type  
0  45,000 kms    Petrol  
1      40 kms    Diesel  
2  22,000 kms    Petrol  
3  28,000 kms    Petrol  
4  36,000 kms    Diesel

### cheking the info of the data

In [4]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 892 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        892 non-null    object
 1   company     892 non-null    object
 2   year        892 non-null    object
 3   Price       892 non-null    object
 4   kms_driven  840 non-null    object
 5   fuel_type   837 non-null    object
dtypes: object(6)
memory usage: 41.9+ KB


*Observations*
* All the variables are in Dtype object 
* Total there are 892 entries and 6 coloumns
* There are missing values in km_driven, Fuel_type

### Checking data of each column

In [5]:
cars['company'].unique()

array(['Hyundai', 'Mahindra', 'Maruti', 'Ford', 'Skoda', 'Audi', 'Toyota',
       'Renault', 'Honda', 'Datsun', 'Mitsubishi', 'Tata', 'Volkswagen',
       'I', 'Chevrolet', 'Mini', 'BMW', 'Nissan', 'Hindustan', 'Fiat',
       'Commercial', 'MARUTI', 'Force', 'Mercedes', 'Land', 'Yamaha',
       'selling', 'URJENT', 'Swift', 'Used', 'Jaguar', 'Jeep', 'tata',
       'Sale', 'very', 'Volvo', 'i', '2012', 'Well', 'all', '7', '9',
       'scratch', 'urgent', 'sell', 'TATA', 'Any', 'Tara'], dtype=object)

*Observations*
* There are company names like 'i','2012','Used','7','9' 
* Need to discuss with client and fix

In [6]:
cars['year'].unique()

array(['2007', '2006', '2018', '2014', '2015', '2012', '2013', '2016',
       '2010', '2017', '2008', '2011', '2019', '2009', '2005', '2000',
       '...', '150k', 'TOUR', '2003', 'r 15', '2004', 'Zest', '/-Rs',
       'sale', '1995', 'ara)', '2002', 'SELL', '2001', 'tion', 'odel',
       '2 bs', 'arry', 'Eon', 'o...', 'ture', 'emi', 'car', 'able', 'no.',
       'd...', 'SALE', 'digo', 'sell', 'd Ex', 'n...', 'e...', 'D...',
       ', Ac', 'go .', 'k...', 'o c4', 'zire', 'cent', 'Sumo', 'cab',
       't xe', 'EV2', 'r...', 'zest'], dtype=object)

*Observations*
* There are lot of values which doesn't represent the year 
* Need to fix and convert it in to nummerical value

In [7]:

pd.set_option('display.max_rows',500)

In [8]:
cars['Price'].value_counts()

Ask For Price    35
2,50,000         17
3,50,000         14
1,80,000         13
1,30,000         12
4,00,000         12
3,00,000         11
4,50,000         11
1,50,000         11
2,20,000         11
3,75,000         10
90,000           10
2,10,000         10
1,20,000         10
2,70,000          9
95,000            9
1,10,000          9
1,75,000          9
1,35,000          9
85,000            9
1,00,000          9
3,20,000          9
1,60,000          9
3,40,000          8
18,91,111         8
2,40,000          8
1,90,000          8
9,00,000          8
5,00,000          8
1,25,000          8
3,80,000          8
2,25,000          8
2,30,000          8
2,65,000          7
3,90,000          7
4,25,000          7
2,00,000          7
80,000            7
60,000            7
6,50,000          7
2,80,000          7
3,65,000          6
2,75,000          6
3,10,000          6
3,99,999          6
5,50,000          6
10,00,000         6
4,75,000          6
6,00,000          6
4,89,999          6


*Observations*
* There is value 'Ask For Price'
* Price is in object format, it should be int of float

In [9]:
cars['kms_driven'].value_counts()

45,000 kms      30
35,000 kms      30
55,000 kms      25
50,000 kms      23
20,000 kms      22
60,000 kms      20
30,000 kms      19
40,000 kms      18
65,000 kms      16
38,000 kms      16
41,000 kms      16
46,000 kms      16
39,000 kms      15
22,000 kms      14
42,000 kms      12
25,000 kms      12
36,000 kms      12
70,000 kms      12
47,000 kms      11
27,000 kms      10
75,000 kms      10
44,000 kms       9
56,000 kms       9
37,000 kms       9
32,000 kms       8
80,000 kms       8
51,000 kms       8
43,000 kms       8
29,000 kms       7
7,000 kms        7
90,000 kms       7
15,000 kms       7
48,000 kms       7
1,00,000 kms     7
33,000 kms       7
68,000 kms       7
13,000 kms       6
34,000 kms       6
23,000 kms       6
1,04,000 kms     6
31,000 kms       6
14,000 kms       6
9,000 kms        5
62,000 kms       5
52,000 kms       5
97,200 kms       5
28,000 kms       5
19,000 kms       4
16,000 kms       4
64,000 kms       4
72,000 kms       4
49,000 kms       4
59,000 kms  

*Observations*
* Ther is value 'Petrol
* It is in object format and 'kms with data
* Nan values

In [10]:
cars['fuel_type'].unique()

array(['Petrol', 'Diesel', nan, 'LPG'], dtype=object)

*Observations*
* Nan values

In [11]:
cars['name'].unique()   

array(['Hyundai Santro Xing XO eRLX Euro III', 'Mahindra Jeep CL550 MDI',
       'Maruti Suzuki Alto 800 Vxi',
       'Hyundai Grand i10 Magna 1.2 Kappa VTVT',
       'Ford EcoSport Titanium 1.5L TDCi', 'Ford Figo', 'Hyundai Eon',
       'Ford EcoSport Ambiente 1.5L TDCi',
       'Maruti Suzuki Alto K10 VXi AMT', 'Skoda Fabia Classic 1.2 MPI',
       'Maruti Suzuki Stingray VXi', 'Hyundai Elite i20 Magna 1.2',
       'Mahindra Scorpio SLE BS IV', 'Audi A8', 'Audi Q7',
       'Mahindra Scorpio S10', 'Maruti Suzuki Alto 800',
       'Hyundai i20 Sportz 1.2', 'Maruti Suzuki Alto 800 Lx',
       'Maruti Suzuki Vitara Brezza ZDi', 'Maruti Suzuki Alto LX',
       'Mahindra Bolero DI', 'Maruti Suzuki Swift Dzire ZDi',
       'Mahindra Scorpio S10 4WD', 'Maruti Suzuki Swift Vdi BSIII',
       'Maruti Suzuki Wagon R VXi BS III',
       'Maruti Suzuki Wagon R VXi Minor',
       'Toyota Innova 2.0 G 8 STR BS IV', 'Renault Lodgy 85 PS RXL',
       'Skoda Yeti Ambition 2.0 TDI CR 4x2',
       'Maru

*Observations*
* No uniformity in the data 

In [12]:
backup = cars.copy()

### Data cleaning 

In [13]:
#Year
#Selecting only numeric values from yesr by using string method
cars = cars[cars['year'].str.isnumeric()]

In [14]:
#Converting the yesr to int 

cars['year']=cars['year'].astype(int)

In [15]:
#Price

cars=cars[cars['Price']!='Ask For Price']

In [16]:
cars['Price'] =cars['Price'].str.replace(',','').astype(int)

In [17]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 819 entries, 0 to 891
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        819 non-null    object
 1   company     819 non-null    object
 2   year        819 non-null    int32 
 3   Price       819 non-null    int32 
 4   kms_driven  819 non-null    object
 5   fuel_type   816 non-null    object
dtypes: int32(2), object(4)
memory usage: 38.4+ KB


In [18]:
# KM driven
cars['kms_driven']

0        45,000 kms
1            40 kms
3        28,000 kms
4        36,000 kms
6        41,000 kms
           ...     
886    1,32,000 kms
888      27,000 kms
889      40,000 kms
890          Petrol
891          Petrol
Name: kms_driven, Length: 819, dtype: object

In [19]:
#removing the kms and ',' from the data
cars['kms_driven'] =cars['kms_driven'].str.split(' ').str.get(0).str.replace(',','')

In [54]:
cars['kms_driven']

0       45000
1          40
2       28000
3       36000
4       41000
        ...  
811     50000
812     30000
813    132000
814     27000
815     40000
Name: kms_driven, Length: 815, dtype: int32

In [21]:
cars['kms_driven']= cars['kms_driven'].astype(int)

In [22]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 817 entries, 0 to 889
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        817 non-null    object
 1   company     817 non-null    object
 2   year        817 non-null    int32 
 3   Price       817 non-null    int32 
 4   kms_driven  817 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 35.1+ KB


In [23]:
#Fuel type
cars = cars[~cars['fuel_type'].isna()]

In [24]:
cars['name'] =cars['name'].str.split(' ').str.slice(0,3).str.join(' ')

In [25]:
cars.head()

name   company  year   Price  kms_driven fuel_type
0     Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1     Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
3       Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
4  Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
6               Ford Figo      Ford  2012  175000       41000    Diesel

In [26]:
cars.reset_index(drop=True, inplace=True)

In [27]:
cars

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
2         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
4                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
811      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
812          Tata Indica V2      Tata  2009  110000       30000    Diesel
813    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
814            Tata Zest XM      Tata  2018  260000       27000    Diesel
815      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[816 rows x 6 columns]

In [28]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        816 non-null    object
 1   company     816 non-null    object
 2   year        816 non-null    int32 
 3   Price       816 non-null    int32 
 4   kms_driven  816 non-null    int32 
 5   fuel_type   816 non-null    object
dtypes: int32(3), object(3)
memory usage: 28.8+ KB


### Exploratory data analysis

In [29]:
cars.describe(include='all')

name company         year         Price     kms_driven  \
count                   816     816   816.000000  8.160000e+02     816.000000   
unique                  254      25          NaN           NaN            NaN   
top     Maruti Suzuki Swift  Maruti          NaN           NaN            NaN   
freq                     51     221          NaN           NaN            NaN   
mean                    NaN     NaN  2012.444853  4.117176e+05   46275.531863   
std                     NaN     NaN     4.002992  4.751844e+05   34297.428044   
min                     NaN     NaN  1995.000000  3.000000e+04       0.000000   
25%                     NaN     NaN  2010.000000  1.750000e+05   27000.000000   
50%                     NaN     NaN  2013.000000  2.999990e+05   41000.000000   
75%                     NaN     NaN  2015.000000  4.912500e+05   56818.500000   
max                     NaN     NaN  2019.000000  8.500003e+06  400000.000000   

       fuel_type  
count        816  
unique         3  
top       Petrol  
freq         428  
mean         NaN  
std          NaN  
min          NaN  
25%          NaN  
50%          NaN  
75%          NaN  
max          NaN

In [30]:
#Mean and median of the car price showing a major difference , So there is a possiblity of outlier
cars[cars['Price']>6000000]

name   company  year    Price  kms_driven fuel_type
534  Mahindra XUV500 W6  Mahindra  2014  8500003       45000    Diesel

In [31]:
cars = cars.drop(index=534,axis=0)

In [32]:
cars.describe()

year         Price     kms_driven
count   815.000000  8.150000e+02     815.000000
mean   2012.442945  4.017933e+05   46277.096933
std       4.005079  3.815888e+05   34318.459638
min    1995.000000  3.000000e+04       0.000000
25%    2010.000000  1.750000e+05   27000.000000
50%    2013.000000  2.999990e+05   41000.000000
75%    2015.000000  4.900000e+05   56879.000000
max    2019.000000  3.100000e+06  400000.000000

In [33]:
cars

name   company  year   Price  kms_driven fuel_type
0       Hyundai Santro Xing   Hyundai  2007   80000       45000    Petrol
1       Mahindra Jeep CL550  Mahindra  2006  425000          40    Diesel
2         Hyundai Grand i10   Hyundai  2014  325000       28000    Petrol
3    Ford EcoSport Titanium      Ford  2014  575000       36000    Diesel
4                 Ford Figo      Ford  2012  175000       41000    Diesel
..                      ...       ...   ...     ...         ...       ...
811      Maruti Suzuki Ritz    Maruti  2011  270000       50000    Petrol
812          Tata Indica V2      Tata  2009  110000       30000    Diesel
813    Toyota Corolla Altis    Toyota  2009  300000      132000    Petrol
814            Tata Zest XM      Tata  2018  260000       27000    Diesel
815      Mahindra Quanto C8  Mahindra  2013  390000       40000    Diesel

[815 rows x 6 columns]

In [34]:
#storing final data to file
cars.to_csv('Cars_cleaned_final.csv')

### ML Model - Linear Regression

In [35]:
x= cars.drop(['Price'],axis=1)
y=cars['Price']

In [36]:
#importing the ML model and preprocess packages
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.preprocessing import OneHotEncoder

In [37]:
#Test train split 
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2)

In [38]:
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((652, 5), (163, 5), (652,), (163,))

In [39]:
ohe = OneHotEncoder()


In [40]:
ohe.fit(x[['name','company','fuel_type']])

OneHotEncoder()

In [41]:
#Data processing using cloumn tranformer and pipeline
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline

In [42]:
col_trans = make_column_transformer((OneHotEncoder(categories=ohe.categories_),['name','company','fuel_type']),
                              remainder='passthrough')

In [43]:
lr =LinearRegression()

In [44]:
pipe = make_pipeline(col_trans,lr)

In [45]:
pipe.fit(x_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Audi A3 Cabriolet', 'Audi A4 1.8', 'Audi A4 2.0', 'Audi A6 2.0',
       'Audi A8', 'Audi Q3 2.0', 'Audi Q5 2.0', 'Audi Q7', 'BMW 3 Series',
       'BMW 5 Series', 'BMW 7 Series', 'BMW X1', 'BMW X1 sDrive20d',
       'BMW X1 xDrive20d', 'Chevrolet Beat', 'Chevrolet Beat...
                                                                            array(['Audi', 'BMW', 'Chevrolet', 'Datsun', 'Fiat', 'Force', 'Ford',
       'Hindustan', 'Honda', 'Hyundai', 'Jaguar', 'Jeep', 'Land',
       'Mahindra', 'Maruti', 'Mercedes', 'Mini', 'Mitsubishi', 'Nissan',
       'Renault', 'Skoda', 'Tata', 'Toyota', 'Volkswagen', 'Volvo'],
      dtype=object),
                                                                            array(['Diesel', 'LPG', 'Pe

In [46]:
y_pred = pipe.predict(x_test)

In [47]:
r2_score(y_test,y_pred) 

0.5775003719848978

In [48]:
for i in range(10):
    x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=i)
    lr =LinearRegression()
    pipe = make_pipeline(col_trans,lr)
    pipe.fit(x_train,y_train)
    y_pred = pipe.predict(x_test)
    print(f'The r2 score is {r2_score(y_test,y_pred)}, {i} ')
    

The r2 score is 0.7852425416626871, 0 
The r2 score is 0.5125329954390436, 1 
The r2 score is 0.6903320946309455, 2 
The r2 score is 0.5995042932558412, 3 
The r2 score is 0.8538307173340905, 4 
The r2 score is 0.7472189241951475, 5 
The r2 score is 0.6257517128888461, 6 
The r2 score is 0.6859812152508706, 7 
The r2 score is 0.7646095915730756, 8 
The r2 score is 0.73107907342576, 9 


In [49]:
#Getting the random state with max r2 score 

In [50]:
score = []
for i in range(1000):
    x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=i)
    lr =LinearRegression()
    pipe = make_pipeline(col_trans,lr)
    pipe.fit(x_train,y_train)
    y_pred = pipe.predict(x_test)
    score.append(r2_score(y_test,y_pred))

In [51]:
max(score)

0.8897762920578806

In [52]:
np.argmax(score)

661

In [53]:
score[np.argmax(score)]

0.8897762920578806